standalone/pruntime/enclave/src/lib.rs

```diff
@@ -1361,15 +1490,46 @@ fn dispatch_block(input: DispatchBlockReq) -> Result<Value, Value> {
             .expect("ECDH not initizlied"),
     );
     let mut last_block = 0;
-    for block in blocks.iter() {
-        if block.events.is_none() {
-            return Err(error_msg("Event was required"));
+    for block in blocks.into_iter() {
+        let expected_root = local_state
+            .block_hashes
+            .get(0)
+            .ok_or(error_msg("No enough headers to validate the blocks"))?
+            .1;
+
+        let changes = &block.storage_changes;
+        local_state.runtime_state.apply_changes(
+            &changes.main_storage_changes,
+            &changes.child_storage_changes,
+        );
+
+        // TODO.kevin: rollback the state on failure
+        let state_root = local_state.runtime_state.root();
+        if &expected_root != state_root {
+            error!("expected root: {:?}", expected_root);
+            error!("real root: {:?}", state_root);
+            return Err(error_msg("State root mismatch"));
         }
+        info!("New state root: {:?}", state_root);
+
+        let event_storage_key = storage_prefix("System", "Events");
+        let events = local_state
+            .runtime_state
+            .get(&event_storage_key)
+            .ok_or(error_msg("Can not get Events from storage"))?;
 
-        handle_events(&block, &ecdh_privkey, local_state.dev_mode)?;
+        let worker_snapshot = snapshot_online_worker(&local_state.runtime_state)?;
```

Comment:
Online worker data is only used when there's a "NewMiningRound" event in the current block. ([reference](https://github.com/Phala-Network/phala-blockchain/blob/cd151d850caaf2f9d5015300e1b8dd66655c7b7d/standalone/phost/src/chain_client.rs#L126-L138))

I'd suggest to let the system contract itself to extract the data. We don't need to extract it at every block outside the system contract.